### DISCUSSION TITLE
is metric learning on images supposed to be so slow, both cuda & normal? "dnn_metric_learning_on_images_ex.cpp.html"

### DISCUSSION THREAD

**Discussion by M-U-X** [davisking/dlib#2916] [General]
so I'm running the metric learning on images example "dnn_metric_learning_on_images_ex.cpp.html" in one project I'm using the vcpkg's dlip, & in another using cmake to generate a visual studio project making sure I've CUDA enabled, build is release of course, with all types of compiler optimizations enabled for all projects, by far the vcpkg one is the fastest, but even it is just using 5% of my CPU i7 14700k, & it's the same for the cuda enabled compile, I'm not sure of the correct way to modify the program to use CPU or GPU, or to show which device dlip is using, I've only been running it on the provided john's folder in example, hardware is i7 14700k, rtx 3060, 32GB of ddr5, so I'm not sure if it's supposed to be this slow, if this part of the process "metric learning" not parallelizable or what's happening exactly

**Comment by davisking** [comment]
I bet it's not using cuda.  In cmake you can check if `DLIB_USE_CUDA` is set to ON or not.  You will also see in the build settings your compiler is using and files it's compiling that it's compiling cuda files if it's using cuda.  And nvidia-smi will show the process using cuda if it's really using cuda.  There are lots of ways to check.  You can also check if `DLIB_USE_CUDA` is `#defined` in the C++ code.

**Reply by M-U-X** [reply]
tysm, indeed it wasn't, I forgot to define DLIB_USE_CUDA  on the project using the library, I only did for the library project, now it's blazingly fast, any idea why the cpu one doesn't use all of my cores though, or what do I need to modify in the example? I already added more data loaders  
all I can think of is increasing the nnumber of pipes
`dlib::pipe<std::vector<matrix<rgb_pixel>>> qimages(4);
dlib::pipe<std::vector<unsigned long>> qlabels(4);`
& increasing the loaders 
`std::thread data_loader1([data_loader]() { data_loader(1); });
std::thread data_loader2([data_loader]() { data_loader(2); });
std::thread data_loader3([data_loader]() { data_loader(3); });
std::thread data_loader4([data_loader]() { data_loader(4); });
std::thread data_loader5([data_loader]() { data_loader(5); });`

also does the `load_mini_batch(5, 11, rnd, objs, images, labels);` affect it somehow?

**Reply by davisking** [reply]
The CPU code is single threaded unless you link to the intel MKL library (or some other multi-threaded BLAS).

Yeah data loading needs to be fast enough to keep up, what exactly you need to do on that end is problem and hardware dependent.

**Reply by M-U-X** [reply]
tysm again, I made a compile using intel mkl for cpu only, & it's working great, as for the data loading depending on my hardware, is this all I need to do to let's say set the number of threads used for loading to N, change this to `dlib::pipe<std::vector<matrix<rgb_pixel>>> qimages(N); dlib::pipe<std::vector<unsigned long>> qlabels(N);` & add 18 of those? `std::thread data_loader1([data_loader]() { data_loader(1); });` ?

**Reply by davisking** [reply]
The size of the pipe doesn't really matter.  Really you just make more thread objects.
